Add OnceCell and Lazy

b-naber opened the pull request:
This PR adds versions of `std::lazy::OnceCell` and `std::lazy::Lazy` that allow one to initialize these objects with async functions. 

Fixes https://github.com/tokio-rs/tokio/issues/3482

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+    /// a reference to that initialized value is returned
+    ///
+    /// This will deadlock if `f` tries to initialize the cell itself.
+    pub async fn get_or_init<F>(&self, f: F) -> &T
+    where
+        F: Fn() -> Pin<Box<dyn Future<Output = T> + Send>>,
+    {
```
Can you add all the async methods to [`async_send_sync`][1] using the `async_assert_fn!` macro? I would also like to see both the `OnceCell` and `Lazy` structs there using the `assert_value!` macro.

[1]: https://github.com/tokio-rs/tokio/blob/master/tokio/tests/async_send_sync.rs

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+///         ONCE.get_or_init(call_async_fn).await
+///     }).await.unwrap();
+///     assert_eq!(*result1, 2);
+/// }
+pub struct OnceCell<T> {
```
Please double-check that you haven't forgotten this in other doc-tests too.
```suggestion
/// }
/// ```
pub struct OnceCell<T> {
```

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+    }
+}
+
+impl<T> fmt::Debug for OnceCell<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("OnceCell")
+            .field("value", &self.value)
+            .finish()
+    }
+}
```
This could be improved to actually print the value if initialized, and print that it isn't if not.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+pub struct OnceCell<T> {
+    value_set: AtomicBool,
+    value: UnsafeCell<MaybeUninit<T>>,
+    sema: Semaphore,
```
I think it is more readable like this.
```suggestion
    semaphore: Semaphore,
```

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+        self.value_set.store(true, Ordering::Release);
+    }
+
+    unsafe fn get_unchecked(&self) -> &T {
+        &*self.value.with(|ptr| (*ptr).as_ptr())
+    }
+
+    unsafe fn set_value(&self, value: T) {
+        self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+    }
```
This method could totally include the `value_set = true` and closing the semaphore.

Darksonn replied:
The unsafe methods should have a comment explaining when it is ok to call them.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+///
+/// [`OnceCell::get`]: crate::sync::OnceCell::get
+/// [`OnceCell::set`]: crate::sync::OnceCell::set
+#[derive(Debug, PartialEq)]
+pub enum OnceCellError {
+    /// The value in OnceCell was already initialized
+    AlreadyInitialized,
+
+    /// The value in OnceCell was not previously initialized
+    NotInitialized,
+}
```
I would like to split this into two error types, as the caller otherwise ends up with extra unreachable branches if they match on it.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+    }
+}
+
+unsafe impl<T: Sync + Send> Sync for OnceCell<T> {}
+unsafe impl<T: Send> Send for OnceCell<T> {}
```
These are correct, but need a comment explaining why.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        panic!("acquired semaphore after value was already initialized");
```
```suggestion
                        unreachable!("acquired semaphore after value was already initialized");
```

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+                        // we can give out arbitrarily many immutable references
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        panic!("semaphore acquire call failed when value not already intialized");
```
```suggestion
                        unreachable!("semaphore acquire call failed when value not already intialized");
```

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+    /// a reference to that initialized value is returned
+    ///
+    /// This will deadlock if `f` tries to initialize the cell itself.
+    pub async fn get_or_init<F>(&self, f: F) -> &T
+    where
+        F: Fn() -> Pin<Box<dyn Future<Output = T> + Send>>,
```
Can be simplified to
```suggestion
    pub async fn get_or_init<F, Fut>(&self, f: F) -> &T
    where
        F: FnOnce() -> Fut,
        Fut: Future<Output = T>,
```

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,300 @@
+        }
+    }
+
+    fn initialized(&self) -> bool {
+        self.value_set.load(Ordering::Acquire)
+    }
```
This method could be public.

Darksonn commented on tokio/tests/sync_once_cell.rs:
```diff
@@ -0,0 +1,97 @@
+    Box::pin(f())
+}
+
+#[test]
+fn get_or_init() {
```
Always good to see tests!

Darksonn reviewed (commented):
Please don't be overwhelmed by the number of comments! Overall, this looks very good.

b-naber commented:
@Darksonn Addressed your comments. I had some problems with the `async_assert_sync` tests however. I got `size for values of type ... not known at compilation time` errors when supplying function types. But thinking about this, I maybe have to use function pointers there again, let me try that.

b-naber commented:
btw is there some way for cargo not to re-compile all test crates again when running `cargo test`?

MikailBag commented on tokio/src/sync/once_cell.rs:
```diff
@@ -50,11 +49,17 @@ impl<T> Default for OnceCell<T> {
-        fmt.debug_struct("OnceCell")
-            .field("value", &self.value)
-            .finish()
+        if self.initialized() {
```
Maybe use `OnceCell::get()` instead, to avoid unsafe block?

b-naber replied:
Yes I could do that, but `OnceCell::get` calls the unsafe method internally. Is there something bad about having `unsafe` blocks in `Debug` impls?

MikailBag replied:
In my opinion, you should not use unsafe if it does not give you any benefits (less unsafe code - less chance for UB and less work for the reviewer). For example, it is totally fine IMHO to use unsafe code to improve performance. But I don't see any gains arising from this unsafe usage.

Anyway, it's just a minor remark, feel free to ignore.

b-naber replied:
No, I think you're right. Not having `unsafe` code makes the code easier to read. I'll change that, thanks.

b-naber commented:
`async_assert_sync` works now.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,310 @@
+    // acquired
+    unsafe fn set_value(&self, value: T) {
+        self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+        self.set_to_initialized();
```
I think I would probably just inline the `set_to_initialized` call into `set_value`.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,310 @@
+        }
+    }
+
+    /// Sets the value of the OnceCell to the argument value.
+    ///
+    /// If the value of the OnceCell was already set prior to this call, then
+    /// [`AlreadyInitializedError`] is returned
+    ///
+    /// [`AlreadyInitializedError`]: crate::sync::AlreadyInitializedError
+    pub fn set(&self, value: T) -> Result<(), AlreadyInitializedError> {
```
The documentation here should explain what happens if some other thread is currently inside `get_or_init`, waiting for the async method to complete.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,310 @@
+        Err(AlreadyInitializedError)
+    }
+
+    /// Tries to initialize the value of the OnceCell using the async function `f`.
+    /// If the value of the OnceCell had already been initialized prior to this call
+    /// a reference to that initialized value is returned
+    ///
+    /// This will deadlock if `f` tries to initialize the cell itself.
+    pub async fn get_or_init<F, Fut>(&self, f: F) -> &T
```
This documentation is a bit unclear. If the `OnceCell` has not already been initialized, what does it return? (I know the answer, but the doc is unclear.)

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,310 @@
+                        return unsafe { self.get_unchecked() };
+                    } else {
+                        unreachable!(
+                            "semaphore acquire call failed when value not already intialized"
```
This is a bit unclear.
```suggestion
                            "Semaphore closed, but the OnceCell has not been initialized."
```

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,310 @@
+    }
+}
+
+// OnceCell is Send (and Sync) because access to its UnsafeCell
+// is guarded by the semaphore permit and atomic operations
+// on `value_set`, as long as T itself is Send (and Sync)
+unsafe impl<T: Sync + Send> Sync for OnceCell<T> {}
+unsafe impl<T: Send> Send for OnceCell<T> {}
```
This explanation seems incomplete. It doesn't explain why `Send` only requires `T: Send`, but `Sync` requires both `T: Send` and `T: Sync`.

b-naber replied:
I don't know the exact reason for why we only need the `Sync` bound on `T` for `OnceCell<T>` to be Sync, I was assuming that inner generic types of a type for which we want to  implement `Sync` always require both bounds to hold. Can you explain, please?

Darksonn replied:
Well the definition of `Sync` is that a type is `Sync` if and only if it is safe to have immutable references to the value in multiple threads. In our case, we need both restrictions because:

 1. You can use an `&OnceCell<T>` to get an `&T`, so we need `T` to be `Sync`.
 2. Due to the `set` method operating on `&OnceCell<T>`, we can use it to _send_ values across threads too.

Strictly speaking the second requirement would only apply if you had a way to access the value in the `OnceCell` mutably, but it would make sense for us to have an `get_mut(&mut self) -> &mut T` and `into_inner(self) -> T` method anyway, so we do need this requirement.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,310 @@
+///
+/// [`OnceCell::set`]: crate::sync::OnceCell::set
+#[derive(Debug, PartialEq)]
+pub struct AlreadyInitializedError;
```
We should implement `Error` for this struct.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,310 @@
+unsafe impl<T: Sync + Send> Sync for OnceCell<T> {}
+unsafe impl<T: Send> Send for OnceCell<T> {}
+
+/// Error returned from the [`OnceCell::set`] method
```
You are missing the period at the end of lots of sentences.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,310 @@
+///
+/// #[tokio::main]
+/// async fn main() {
+///     static LAZY : Lazy<u32> = Lazy::new(call_async_fn);
```
I don't like having the global inside the main function.

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,310 @@
+///     1 + 1
+/// }
+///
+/// fn call_async_fn() -> Pin<Box<dyn Future<Output = u32> + Send>> {
+///     Box::pin(some_computation())
+/// }
```
I would prefer if this was done with a closure.

Darksonn commented on tokio/src/sync/once_cell.rs:
````diff
@@ -0,0 +1,310 @@
+/// ```
+pub struct Lazy<T, F = fn() -> Pin<Box<dyn Future<Output = T> + Send>>> {
+    cell: OnceCell<T>,
+    init: Cell<Option<F>>,
````
Using a `Cell` in a `Sync` struct feels somewhat like we are hiding some unsafe pieces. I would prefer if this was an `UnsafeCell` so the compiler makes us actually annotate the parts of the code that do unsafe things.

Darksonn commented on tokio/tests/async_send_sync.rs:
```diff
@@ -265,6 +267,17 @@ async_assert_fn!(tokio::sync::watch::Sender<u8>::closed(_): Send & Sync);
 async_assert_fn!(tokio::sync::watch::Sender<Cell<u8>>::closed(_): !Send & !Sync);
 async_assert_fn!(tokio::sync::watch::Sender<Rc<u8>>::closed(_): !Send & !Sync);
 
+async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
+    _, fn() -> Pin<Box<dyn Future<Output = u8> + Send>>): Send & !Sync);
```
Why is it not `Sync`? It's a future — those can always be `Sync`.

b-naber replied:
Why does the assertion fail if I use `Sync`?

```
error[E0277]: `dyn Future<Output = u8> + Send` cannot be shared between threads safely
   --> tokio/tests/async_send_sync.rs:114:26
    |
30  |   fn require_sync<T: Sync>(_t: &T) {}
    |                      ---- required by this bound in `require_sync`
...
114 |               require_sync(&f);
    |                            ^^ `dyn Future<Output = u8> + Send` cannot be shared between threads safely
...
270 | / async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
271 | |     _, fn() -> Pin<Box<dyn Future<Output = u8> + Send>>): Send & Sync);
```

Darksonn replied:
The assertion fails because the future returned by the method is not `Sync`. Mostly I'm thinking that it _should_ be `Sync`.

b-naber replied:
The tests only pass in its current form.

Darksonn replied:
Yes, I know that. My point is that there is likely some change you could make to `get_or_init` that _would_ make the test pass with `Sync`.

b-naber replied:
Why do we want the method to be `Sync`? Isn't it sufficient that OnceCell is `Sync`? Once we got the reference to an `OnceCell` instance we can still call `get_or_init`, can`t we?

Darksonn replied:
Generally it is more convenient for the user if the things implement as many traits as possible. For example, it not being `Sync` would prevent you from putting an async fn that calls `get_or_init` into an `Box<dyn Future + Send + Sync>`, even though there is no safety-concern with doing so.

Darksonn replied:
Thinking more about it, it is fine if we have both of these tests:
```Rust
async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
    _, fn() -> Pin<Box<dyn Future<Output = u8> + Send>>): Send & !Sync);
async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
    _, fn() -> Pin<Box<dyn Future<Output = u8> + Send + Sync>>): Send & Sync);
```

b-naber replied:
> Generally it is more convenient for the user if the things implement as many traits as possible. For example, it not being `Sync` would prevent you from putting an async fn that calls `get_or_init` into an `Box<dyn Future + Send + Sync>`, even though there is no safety-concern with doing so.

That makes sense in general, but in our case the following works e.g.:

```
use std::future::Future;
use std::pin::Pin;
use tokio;
use tokio::sync::OnceCell;

static ONCE: OnceCell<u32> = OnceCell::new();

async fn some_computation() -> u32 {
    5
}

async fn call_get_or_init() -> u32 {
    *ONCE.get_or_init(|| some_computation()).await
}

async fn call_f(f: Pin<Box<dyn Future<Output = u32> + Send + Sync>>) {
    f.await;
}

#[tokio::main]
async fn main() {
    tokio::spawn(async {
        call_f(Box::pin(call_get_or_init())).await;
    });
}
```

Darksonn replied:
Yes. I realized it was because the future type itself is not `Sync` in the test. Your snippet is in the _second_ case of my two test lines.

Darksonn commented on tokio/tests/async_send_sync.rs:
```diff
@@ -265,6 +267,17 @@ async_assert_fn!(tokio::sync::watch::Sender<u8>::closed(_): Send & Sync);
+    _, fn() -> Pin<Box<dyn Future<Output = Cell<u8>>>>): !Send & !Sync);
+async_assert_fn!(tokio::sync::Lazy<u8>::get(_): Send & !Sync);
+async_assert_fn!(tokio::sync::Lazy<Cell<u8>>::get(_): !Send & !Sync);
+assert_value!(tokio::sync::OnceCell<u8>: Send & Sync);
+assert_value!(tokio::sync::OnceCell<Rc<u8>>: !Send & !Sync);
```
We need to go through all of `u8`, `Cell<u8>` and `Rc<u8>` here to check all three cases:

 1. `Send & Sync`
 2. `Send & !Sync`
 3. `!Send & !Sync`

In fact, you are missing the most interesting case!

Darksonn commented on tokio/tests/sync_once_cell.rs:
```diff
@@ -0,0 +1,97 @@
+    static ONCE: OnceCell<u32> = OnceCell::new();
+
+    rt.block_on(async {
+        let result1 = rt
+            .spawn(async { ONCE.get_or_init(func1).await })
+            .await
+            .unwrap();
+
+        let result2 = rt
+            .spawn(async { ONCE.get_or_init(func2).await })
+            .await
+            .unwrap();
```
I'm missing a test that does this, but where the second call uses a method that panics.

Darksonn commented on tokio/tests/sync_once_cell.rs:
```diff
@@ -0,0 +1,97 @@
+fn lazy() {
+    // func2 sleeps for 1 second
+    static LAZY: Lazy<u32> = Lazy::new(func2);
+    let rt = Runtime::new().unwrap();
```
Can you use the pause time feature to make this test run faster?

b-naber replied:
I'm not sure whether I misunderstand you here. But if we pause the timer, how is the test supposed to pass?

Darksonn replied:
Afaik the timer automatically advances if things are only waiting for timers. Alternatively please change the durations to something like 10ms so it doesn't take a full second to run them.

b-naber replied:
Yes, 1 second was a conservative estimate. Could the time the runtime takes to execute (or more precisely if the second task is executed before the first one) not lead to the task failing?

Darksonn replied:
Generally, avoiding this kind of thing is the purpose of the time-pausing utilities.

b-naber replied:
It wasn't clear to me how to use `time::pause` to avoid this problem. `time::pause` isn't used in any tests in the code base besides those that test the timer functionality itself.

Johannesd3 commented:
Is there a reason why `get_or_init` takes a `FnOnce() -> impl Future` instead of the future directly?

b-naber commented:
> Is there a reason why `get_or_init` takes a `FnOnce() -> impl Future` instead of the future directly?

Well one reason is that you requested this in the issue. Another is that we can pass async functions to `get_or_init` this way.

Johannesd3 commented:
> > Is there a reason why `get_or_init` takes a `FnOnce() -> impl Future` instead of the future directly?
>
> Well one reason is that you requested this in the issue. Another is that we can pass async functions to `get_or_init` this way.

I did not think about this so much. But if you could pass async fns by just calling this function without awaiting it, as in `tokio::spawn`.

b-naber commented:
> > > Is there a reason why `get_or_init` takes a `FnOnce() -> impl Future` instead of the future directly?
> > 
> > 
> > Well one reason is that you requested this in the issue. Another is that we can pass async functions to `get_or_init` this way.
> 
> I did not think about this so much. But you could always pass async fns by just calling this function without awaiting it, as in `tokio::spawn`.

Yes, but you can also pass any future to `get_or_init` by using a closure, I don't think it makes that much difference.

MikailBag commented:
An advantage of the proposed `get_or_init` API is that it does not create a future if there is no need to do so.
However, it _seems_ to be quite minor (future creation usually is very cheap).

Therefore, I'd propose to provide two functions (naming TBD):

```rust
fn get_or_init(&self, f: impl Future);
fn get_or_init_with(&self, f: impl Fn() -> impl Future);
```

That way for usual uses API will be a tiny bit nicer.

Darksonn commented:
I mean, an async block could be used to make it lazy, just like what a closure would do. I would be ok with just having a method that takes a future.

b-naber commented:
@Darksonn I'd prefer not to change this unless you insist on this change being made. I could add an example for how to call futures directly using `get_or_init` and `Lazy`.

Darksonn commented:
I'm fine either way.

b-naber commented:
@Darksonn I think I have addressed all your requests, is there still something you want to be changed?

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,324 @@
+            match self.semaphore.acquire().await {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        let value = f().await;
```
Might be good to mention here that we might be aborted here due to a panic or `select!`.

Darksonn commented on tokio/src/sync/once_cell.rs:
````diff
@@ -0,0 +1,324 @@
+/// ```
+pub struct Lazy<T, F = fn() -> Pin<Box<dyn Future<Output = T> + Send>>> {
+    cell: OnceCell<T>,
+    init: UnsafeCell<Option<F>>,
````
I don't understand why this is not just
```suggestion
    init: F,
```

Johannesd3 replied:
If `F` was `FnOnce`, it would be necessary. Is there a reason why `F` must be called more than once?

Johannesd3 replied:
Related: Is it correct that currently, if one future is dropped (one task isn't interested in the result anymore), the next computation will start again from the beginning?

Darksonn replied:
It must be `Fn` in case the operation is cancelled. And to the other question, yes as well.

Johannesd3 replied:
And there's no other way? Something like storing the future inside of the Lazy, so that it can be continued even if the first caller is dropped?

Darksonn replied:
Well maybe .. but you would need a full unsafe implementation for it similar to `OnceCell`. And `OnceCell` would not be able to have the same feature.

Johannesd3 replied:
Unless we modify the implementation of OnceCell as well: 
* Store a `Future<Output = Result<_,_>>` inside the `OnceCell`.
* Insert the `Future` of `get_or_init` only if there's no `Future` present.
* The first waiting task tries to drive the future to completion.
* If the task is dropped, the next one continues the work on the same `Future`.
* If the `Future` completes with an `Err`, it is removed from the `OnceCell`, and the next available `Future` will be inserted.
* If the `Future` completes with an `Ok`, we are ready.
* There is a function to join the waiting for the result without having to provide an init fn.

But this would be yet harder to implement (if possible at all?)

Darksonn commented on tokio/src/sync/once_cell.rs:
```diff
@@ -0,0 +1,324 @@
+    }
+}
+
+// We never create a `&F` from a `&Lazy<T, F>` so it is fine
+// to not require a `Sync` bound on `F`
+unsafe impl<T, F: Send> Sync for Lazy<T, F> where OnceCell<T>: Sync {}
```
This is not true. You do this inside `get`. You can remove both impls, since if you remove the unnecessary `UnsafeCell`, you don't need any unsafe to implement `Lazy`.

Darksonn commented on tokio/tests/async_send_sync.rs:
```diff
@@ -265,6 +267,22 @@ async_assert_fn!(tokio::sync::watch::Sender<u8>::closed(_): Send & Sync);
 async_assert_fn!(tokio::sync::watch::Sender<Cell<u8>>::closed(_): !Send & !Sync);
 async_assert_fn!(tokio::sync::watch::Sender<Rc<u8>>::closed(_): !Send & !Sync);
 
+async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
+    _, fn() -> Pin<Box<dyn Future<Output = u8> + Send + Sync>>): Send & Sync);
+async_assert_fn!(tokio::sync::OnceCell<u8>::get_or_init(
+    _, fn() -> Pin<Box<dyn Future<Output = u8> + Send>>): Send & !Sync);
+async_assert_fn!(tokio::sync::OnceCell<Rc<u8>>::get_or_init(
+    _, fn() -> Pin<Box<dyn Future<Output = Rc<u8>>>>): !Send & !Sync);
```
We are missing a bunch of combinations here.

Darksonn commented:
And also, the tests need to pass.

Darksonn commented:
This PR has had a lot of discussion on the API of the type. I would prefer if we open two new issues:

 1. An issue for discussing the design of `OnceCell`.
 2. An issue for discussing the design of `Lazy`.

The issues should start out with a suggested list of functions that the primitives should have. Then after that, it would probably be good to split it into two PRs, as it has proved a large change.

b-naber commented:
That's fine with me. I also felt like this was not properly specified to begin with. I'd still be interested in implementing the resulting API.

Darksonn commented:
Alright. In that case, please go ahead and open two new issues, each specifically intended for designing the API of each.

Johannesd3 commented:
Or maybe it's a good idea to create an own crate to try out different implementations? If we come to a decision which implementation is the best, we can add it to tokio.

b-naber commented:
After having read the comments again, there was really only one aspect of the API design that was discussed, that being whether `OnceCell::get_or_init` should take an async function or a Future. I fail to see what exactly you would want to discuss about the API design in these issues, let alone why this would need two separate PRs, @Darksonn.

Darksonn commented:
The reason I asked for it is that I find the history here overwhelmingly large, which makes the PR difficult to review. Even if there is consensus on the API, I would still like to split the PR into two.

b-naber commented:
I understand. If you have more than that one proposed API change, it would make sense to me to open an issue to discuss the suggestions, but given only this one discussion point, it would make more sense to me to just split this PR into two and maybe just implement both suggested `get_or_init` methods?

Darksonn commented:
There are some other methods it would make sense to add too (when you have a `&mut` to the `OnceCell`/`Lazy` you can do additional things), but it doesn't have to be part of the initial PR. I'm ok with just splitting the PR without creating issues.

Darksonn commented:
Since #3591 was merged, I will close this. If you want to also add `Lazy`, you should open a new PR.
